delegate-app/Main.hs

```diff
@@ -5,45 +5,57 @@ import Prelude
 
 -- Haskell imports
 
-import Control.Concurrent (threadDelay)
+import Control.Concurrent (Chan, newChan, readChan, threadDelay, writeChan)
+import Control.Concurrent.Async (async)
+import Control.Monad (forever, void)
 import Control.Monad.IO.Class (MonadIO (liftIO))
 
+-- Plutus imports
+import Plutus.V2.Ledger.Api (POSIXTime (..))
+
 -- HydraAuction imports
+
 import HydraAuction.Delegate (
-  DelegateEvent (Start),
+  DelegateEvent (..),
   DelegateInput (DelegateEvent),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
-
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
-eventsProducer thisIsFirstRun =
-  if thisIsFirstRun
-    then return $ Just Start
-    else return Nothing
-
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
-
-delegateTick :: Int
-delegateTick = 1_000
+import HydraAuction.Delegate.Interface (DelegateResponse (..))
+import HydraAuction.OnChain.Common (secondsLeftInInterval, stageToInterval)
+import HydraAuction.Tx.Common (currentAuctionStage, currentTimeMilliseconds)
+
+consumer :: Chan DelegateInput -> DelegateRunnerT IO ()
+consumer delegateInputs = do
+  -- This will block until an event appears in the queue
+  delegateIn <- liftIO $ readChan delegateInputs
+  delegateResponse <- delegateStep delegateIn
+  liftIO $ mbQueueAuctionPhases delegateResponse delegateInputs
+  -- FIXME: send response to client
+  liftIO $
+    putStrLn $
+      "Delegate responses for input: " <> show delegateResponse
+
+mbQueueAuctionPhases :: [DelegateResponse] -> Chan DelegateInput -> IO ()
```

Comment:
Why did you decide for Chan instead of TChan?

Reply:
No real reason, we don't really need complex atomic blocks (yet). Feel free to upgrade to a TChan if needed